scripts/remotes.lua

```diff
@@ -162,86 +366,104 @@ function remotes.notify_player(player, message, is_error)
---
-function remotes.optimise_targeting(targets, damage_radius)
-
-  local optimised_targets = {}
-
-  -- Calculates new target position that is positioned in-between the passed-in targets.
-  local function merge_targets(target1, target2)
-    local new_x = target1.x - math.floor((target1.x - target2.x) / 2)
-    local new_y = target1.y - math.floor((target1.y - target2.y) / 2)
-
-    return {x = new_x, y = new_y}
+-- @return updated array of clusters
+function assign_clusters(targets, radius, clusters)
+  if center_shift == nil then
+    center_shift = false
   end
-
-  -- Checks if two targets could be merged into single one positioned midway between the two.
-  local function can_merge_targets(target1, target2, damage_radius)
-    local x, y = target1.x - target2.x, target1.y - target2.y
-
-    return (x^2 + y^2 < (damage_radius * 2)^2)
-  end
-
-  -- Keep track of processed targets - those have already been optimised and cannot be dropped without risking the full
-  -- explosion coverage. Should map keys to boolean values.
-  local processed = {}
-
-  -- Iterate over all targets and see if it is possible to merge them in pairs.
-  for i1, target1 in pairs(targets) do
-    if not processed[i1] then
-
-      -- Assume target is already optimised (this value is used in one of the later steps).
-      local optimised_target = target1
-
-      -- Merge the target with another one if possible.
-      for i2, target2 in pairs(targets) do
-
-        if i1 ~= i2 and not processed[i1] and not processed[i2] and can_merge_targets(target1, target2, damage_radius) then
-          optimised_target = merge_targets(target1, target2)
-          processed[i2] = true
-          break
+  -- assign targets to clusters
+  for _, target in pairs(targets) do
+    if #clusters == 0 then
+      table.insert(clusters, { center = shape_center(target), radius = radius, targets = { target } })
+    else
+      local best_cluster = nil
+      local best_result = nil
+      for _, cluster in pairs(clusters) do
+        -- try add
+        local try_add_result = try_add_to_cluster(cluster, target)
+        if try_add_result.success and (not best_result or try_add_result.distance < best_result.distance) then
+          best_cluster = cluster
+          best_result = try_add_result
         end
-
       end
-
-      -- Mark the starting target as processed (at this point it is going to be kept, or replaced with a merged target).
-      processed[i1] = true
-
-      -- Check if any other (unprocessed) targets fall within the damage radius of optimised target.
-      for i, target in pairs(targets) do
-
-        -- Reuse the test, pass-in half the damage radius since we don't want to check the mid-point.
-        if not processed[i] and can_merge_targets(optimised_target, target, damage_radius/2) then
-          processed[i] = true
-        end
-
+      if best_cluster then
+        add_to_cluster(best_cluster, target, best_result)
```

Comment:
This was just to make the code a bit easier to read and limit nesting and local complexity of this function. Basically readability and maintainability stuff.

Reply:
Yeah, but as the code is, all you have to do is do table.insert and update the best_cluster.center? Since best_cluster is already passing the conditional from `add_to_cluster` function anyway (it will always be valid based on previous code), and it has no other uses elsewhere in the code.